tools/generate.py

```diff
@@ -1,276 +1,404 @@
+        """
+        self._fp = db_path
+
+    def count(
+        self,
+        sentences_detailed_path: Path,
+        sentence_blacklist: list,
+        buffer_length: int = 5000000,
+    ) -> None:
+        """Count n-grams' occurrences for every Tatoeba language
+
+        A contribution score equal to sum(len(sentences))
+        is also computed for every language a user has contributed to.
+
+        All results are gradually saved into database tables in
+        order to mitigate the memory footprint of the process.
+
+        Parameters
+        ----------
+        sentences_detailed_path : Path
+            path of the 'sentences_detailed.csv' weekly dump file
+        sentence_blacklist : list
+            the integer ids of the sentences that are not taken
+            into account for this counting
+        buffer_length : int, optional
+            the maximum number of n-grams for which counts are kept in RAM,
+            by default 5,000,000.
+            Note that process speed mainly depends on the amount of data
+            written to disk and consequently increases with the buffer size.
+        """
+        # the former database is overwritten
+        if self._fp.exists():
+            print(f"Deleting former {self._fp.name} database")
+            self._fp.unlink()
+        # initialize tables
+        self._init_db()
+
+        user_lang_score = defaultdict(int)
+        for n in range(MAX_NGRAM_SIZE, MIN_NGRAM_SIZE - 1, -1):
+            table_name = f"grams{n}"
+            lang_ngram_cnt = defaultdict(lambda: defaultdict(int))
+            with open(sentences_detailed_path, "r", encoding="utf-8") as f:
+                for line_id, line in enumerate(f):
+                    self._print_status(line_id, ngram_size=n)
+                    try:
+                        fields = line.rstrip().split("\t")
+                        sent_id, lang, text, user = fields[:4]
+                    except IndexError:
+                        print(f"Skipped erroneous line {line_id}: {line}")
+                        continue
+
+                    # sentences with an unset language are ignored
+                    if lang == "\\N" or lang == "":
+                        continue
+                    # sentences with an id matching the blacklist are ignored
+                    if int(sent_id) in sentence_blacklist:
+                        continue
+
+                    # compute user contribution score during last file reading
+                    if n == MIN_NGRAM_SIZE:
+                        user_lang_score[(user, lang)] += len(text)
+
+                    # increment hit counts for each ngram in the sentence
+                    for i in range(len(text) - n + 1):
+                        ngram = text[i : i + n]
+                        lang_ngram_cnt[lang][ngram] += 1
+
+                    # when buffer is full, save it into table and empty it
+                    tot_ngrams = sum(len(v) for v in lang_ngram_cnt.values())
```

Comment:
Isn’t it a bit expensive to compute `tot_ngrams` on every single loop? I am not so familiar with Python so I don’t know how quick is this, but I feel like the "buffering algorithm" could be improved somehow.

For lang_ngram_cnt, what if you use a single dictionary with (lang, ngram) as keys instead of a dictionary of dictionaries?

Also, how about using sys.getsizeof() to evaluate the actual RAM usage (as opposed to the number of items?

Reply:
I think that the length value of a Python object is stored in the object itself and does not have to be calculated when it is called. This is why getting the length of a Python instance is very fast. But I admit we could avoid to re-sum at each loop, even if what mostly slows down the script is the repeated writing of data to disk.

I didn't use `getsizeof ` because it only accounts for the memory consumption directly attributed to an object, not the memory consumption of objects it refers to. I am more comfortable with `len`, but I admit evaluating the actual RAM is more appealing. 

I'll reconsider the `len` choice and set up a "nested dict vs dict with tuple keys" benchmark.

Reply:
I now track the peak memory usage of the process and correlate it with the number of ngrams in the buffer at flush time. Tracking the memory usage in real time with `tracemalloc`  would allow a stricter compliance with the memory limit but would also significantly affect speed.

Reply:
For `lang_ngram_cnt`, the "dict with tuple key" alternative is about 40% slower than the current "nested dicts" implementation.
The nested dicts stay.